src/DeliveryCartReferencedContent.php

```diff
@@ -0,0 +1,189 @@
+  public static function referenceContentHook(EntityInterface $entity) {
+    \Drupal::moduleHandler()
+      ->invokeAll('delivery_cart_referenced_content', [$entity]);
+  }
+
+  /**
+   * Looks for any referenced blocks in layout builder and adds them to the
+   * cart.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *
+   * @return false|int
+   */
+  public static function addBlocksFromLayoutBuilder(EntityInterface $entity) {
+    if (!self::hasField($entity, 'layout_builder__layout')) {
+      return FALSE;
+    }
+
+    $sections = $entity->get('layout_builder__layout')->getSections();
+
+    foreach ($sections as $section) {
+      $components = $section->getComponents();
+      foreach ($components as $component) {
+        $renderedArray = $component->toRenderArray();
+        $theme = $renderedArray['#theme'] ? $renderedArray['#theme'] : '';
+        if ($theme === "block") {
+          $content = $renderedArray['content'] ? $renderedArray['content'] : '';
+          if (isset($content['#block_content'])) {
+            $blockContent = $content['#block_content'];
+            \Drupal::service('delivery.cart')->addToCart($blockContent);
+            self::$count++;
+          }
+        }
+      }
+    }
+
+    return self::$count;
+  }
+
+  /**
+   * Adds referenced nodes.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   */
+  public static function addReferencedNodes(EntityInterface $entity) {
+    if ($entity->getEntityTypeId() === "node") {
+      // Extract all fields and look for the ones using sections.
+      $fieldsWithReferences = [];
+      $fields = $entity->getFieldDefinitions();
+      if (!empty($fields)) {
+        foreach ($fields as $field) {
+          $type = $field->getType();
+          if ($type === "entity_reference" && $field instanceof \Drupal\field\Entity\FieldConfig) {
+            $entityTarget = $field->getTargetEntityTypeId();
+            if ($entityTarget === "node") {
+              $fieldsWithReferences[$field->getName()] = [
+                'field' => $field->getName(),
+                'entityTarget' => $entityTarget,
+              ];
+            }
+          }
+        }
+      }
+
+      if (!empty($fieldsWithReferences)) {
+        foreach ($fieldsWithReferences as $data) {
+          $name = $data['field'];
+          $entityTarget = $data['entityTarget'];
+          if ($entity->hasField($name)) {
+            $values = $entity->get($name)->getValue();
+            if (!empty($values)) {
+
+              foreach ($values as $value) {
+                $targetId = $value['target_id'];
+                $loadEntity = \Drupal::entityTypeManager()
+                  ->getStorage($entityTarget)
+                  ->load($targetId);
+
+                if ($loadEntity) {
+                  \Drupal::service('delivery.cart')->addToCart($loadEntity);
+                  self::$count++;
+                  self::findAllRelatedContent($loadEntity);
+                }
+
```

Comment:
Hi @elistone !
Couldn't it we just add the entities to the cart in one single loop? Whey do we exactly need the $fieldsWithReferences variable?

Reply:
No real reason, was just trying to make the code easier to read. Thought that if first loop was extracting the information required and the second would process that it be a little easier to understand what is going on.

I am happy to combine this still if you want?

Reply:
Have combined it :)